keps/0007-dynamic-audit-control.md

```diff
@@ -0,0 +1,126 @@
+
+### Non-Goals
+- Provide a generic interface to configure all kube-apiserver flags
+
+## Proposal
+
+### Dynamic Configuration of Policy
+Dynamic configuration of the existing policy object will be provided through use of a shared informer.
+```yaml
+apiVersion: audit.k8s.io/v1beta1
+kind: Policy
+omitStages:
+  - "RequestReceived"
+rules:
+  - level: <request level>
+    resources:
+    - group: <group>
+      resources: <resources>
+```
+In kube-apiserver today this is only accepted as a file, this KEP would allow it to either be configured as a file on the master node or applied to the `kube-system` namespace as an object and be dynamically configured. This will require the audit API to be registered with the apiserver. If both file and object exist the dynamic object will take precedence.
+
+### Dynamic Configuration of Flags
+There are a number of configuration options that are currently set through runtime flags. This KEP would provide a new API for the dynamic configuration of the webhook parameters.
+```yaml
+apiVersion: audit.k8s.io/v1alpha1
+kind: AuditConfiguration
+metadata:
+  name: <name>
+webhook:
+  mode: <batch or blocking>
+  initialBackoff: <10s>
+  batch:
+    bufferSize: <400>
+    maxSize: <10000>
+    maxWait: <30s>
+    throttleBurst: <15>
+    throttleEnabled: <true>
+    throttleQPS: <10>
+```
+The webhook would share the model of validating and mutating admission control plugins, leveraging a shared informer to provide online configuration. If runtime flags are also provided, the dynamic configuration will override any parameters provided there.
+
+### User Stories
+
+#### Story 1
+As a cluster admin I will easily be able to enable the interal auditing features of an existing cluster, and tweak the configurations as necessary.
+
+#### Story 2
+As a Kubernetes extension developer, I will be able to provide drop in extensions that utilize audit data.
+
+### Implementation Details/Notes/Constraints
+
+A misunderstanding may occur of both flags are defined and an object is provided. This should be explicitly called out in the 
+documentation and monitored within the community. If this becomes a point of contention we will need to reevaluate the implementation.
+
+### Risks and Mitigations
+
+This does open up the attack surface of the audit mechanisms. Having them strictly configured through the api server has 
+the advantage of limiting the access of those configurations to those that have access to the master node. However, Dynamic Admission has shown that this can be done in a safe manner. 
+
+## Graduation Criteria
+
+Sucess will be determined by stability of the provided mechanisms and ease of understanding for the end user.
+
+* alpha: Policy and api server flags can be dynamically configured, known issues are tested and resolved.
+* beta: Mechanisms have been hardened against any known bugs and the process is validated by the community
+
+## Implementation History
+
+- 05/18/2018: initial design
+
+## Alternatives
```

Comment:
I like this idea, I think inherently you'll want stricter policy on certain namespaces while others cause bloat. We'll track this as a later feature

Reply:
> As an extension, do you think it would ever make sense to allow per-namespace policies

This is what my boss want. I'am happy to implement it.
I think a quota for each namespace's webhooks is necessary.